services/device-registry-mongodb/src/main/java/org/eclipse/hono/deviceregistry/mongodb/service/MongoDbBasedRegistrationService.java

```diff
@@ -0,0 +1,369 @@
+    private MongoDbBasedRegistrationConfigProperties config;
+    private MongoDbCallExecutor mongoDbCallExecutor;
+
+    /**
+     * Registration service, based on {@link AbstractRegistrationService}.
+     * <p>
+     * This helps work around Java's inability to inherit from multiple base classes. We create a new Registration
+     * service, overriding the implementation of {@link AbstractRegistrationService} with the implementation of our
+     * {@link MongoDbBasedRegistrationService#getDevice(String, String, Span)}.
+     */
+    private final AbstractRegistrationService registrationService = new AbstractRegistrationService() {
+
+        @Override
+        public Future<RegistrationResult> getDevice(final String tenantId, final String deviceId, final Span span) {
+            return MongoDbBasedRegistrationService.this.getDevice(tenantId, deviceId, span);
+        }
+
+        @Override
+        public Future<JsonArray> resolveGroupMembers(final String tenantId, final JsonArray viaGroups, final Span span) {
+            return MongoDbBasedRegistrationService.this.resolveGroupMembers(tenantId, viaGroups);
+        }
+    };
+
+    /**
+     * Creates an instance of the {@link MongoDbCallExecutor}.
+     *
+     * @param mongoDbCallExecutor An instance of the mongoDbCallExecutor.
+     */
+    @Autowired
+    public void setExecutor(final MongoDbCallExecutor mongoDbCallExecutor) {
+        this.mongoDbCallExecutor = mongoDbCallExecutor;
+        this.mongoClient = this.mongoDbCallExecutor.getMongoClient();
+    }
+
+    @Autowired
+    public void setConfig(final MongoDbBasedRegistrationConfigProperties config) {
+        this.config = config;
+    }
+
+    public MongoDbBasedRegistrationConfigProperties getConfig() {
+        return config;
+    }
+
+    @Override
+    public void start(final Promise<Void> startPromise) {
+
+        mongoDbCallExecutor.createCollectionIndex(getConfig().getCollectionName(),
+                new JsonObject().put(RegistrationConstants.FIELD_PAYLOAD_TENANT_ID, 1)
+                        .put(RegistrationConstants.FIELD_PAYLOAD_DEVICE_ID, 1),
+                new IndexOptions().unique(true))
+                .map(success -> {
+                    startPromise.complete();
+                    return null;
+                })
+                .onFailure(error -> {
+                    log.error("Index creation failed", error);
+                    startPromise.fail(error);
+                });
+    }
+
+    @Override
+    public void stop(final Promise<Void> stopPromise) {
+        mongoClient.close();
+        stopPromise.complete();
+    }
+
+    @Override
+    public Future<OperationResult<Id>> createDevice(final String tenantId, final Optional<String> deviceId,
+            final Device device, final Span span) {
+
+        Objects.requireNonNull(tenantId);
+        Objects.requireNonNull(deviceId);
+
+        final String deviceIdValue = deviceId.orElse(UUID.randomUUID().toString());
```

Comment:
how about encapsulating the generating of (unique) identifiers in a utility method so that the strategy can be implemented/modified in a single place?